experimental/bakeable/src/lib.rs

```diff
@@ -0,0 +1,176 @@
+extern crate alloc;
+use alloc::borrow::Cow;
+use proc_macro2::TokenStream;
+use quote::quote;
+
+#[cfg(feature = "derive")]
+pub use bakeable_derive::Bakeable;
+
+pub trait Bakeable {
+    fn bake(&self) -> TokenStream;
+}
+
+macro_rules! literal {
+    ($($type:ty),*) => {
+        $(
+            impl Bakeable for $type {
+                fn bake(&self) -> TokenStream {
+                    quote! {
+                        #self
+                    }
+                }
+            }
+        )*
+    }
+}
+
+literal!(u8, u16, u32, u64, u128, usize, i8, i16, i32, i64, i128, isize, &str, str);
+
+impl<T> Bakeable for [T]
+where
+    T: Bakeable,
+{
+    fn bake(&self) -> TokenStream {
+        let data = self.iter().map(|d| d.bake());
+        quote! {
+            &[#(#data),*]
+        }
+    }
+}
+
+impl<T> Bakeable for Option<T>
+where
+    T: Bakeable,
+{
+    fn bake(&self) -> TokenStream {
+        match self {
+            None => quote! { None },
+            Some(t) => {
+                let t = t.bake();
+                quote! {
+                    Some(#t)
+                }
+            }
+        }
+    }
+}
+
+macro_rules! tuple {
+    ($($ty:ident, $ident:ident),*) => {
+        impl<$($ty),*> Bakeable for ($($ty),*) where $($ty: Bakeable),* {
+            fn bake(&self) -> TokenStream {
+                let ($($ident),*) = self;
+                $(
+                    let $ident = $ident.bake();
+                )*
+                quote! {
+                    ($(#$ident),*)
+                }
+            }
+        }
+    }
+}
+
+tuple!(A, a, B, b);
+tuple!(A, a, B, b, C, c);
+tuple!(A, a, B, b, C, c, D, d);
+tuple!(A, a, B, b, C, c, D, d, E, e);
+
+macro_rules! cow {
+    ($($ty:ty),*) => {
+        $(
+            impl<'a> Bakeable for Cow<'a, $ty> {
```

Comment:
suggestion: i think this can be generic using `impl<T> Bakeable for Cow<'_, T> where &T: Bakeable`

Reply:
Can't figure out the lifetime bounds for this

Reply:
I think it needs to be `impl<'a, T> Bakeable for Cow<'a, T> where &'static T: Bakeable, T: 'static`

paired with `impl<T: Bakeable> Bakeable for &'static T`

Reply:
Nope. Could you give it a try?

Reply:
Done, pushed

Reply:
ah yes, `&&**` how could I miss that :D